third_party/move/move-model/src/builder/builtins.rs

```diff
@@ -89,49 +97,241 @@ pub(crate) fn declare_builtins(trans: &mut ModelBuilder<'_>) {
+                           visibility: EntryVisibility| {
+            declare_bin_gen(
+                trans,
+                op,
+                oper,
+                &[],
+                &BTreeMap::default(),
+                param_type1,
+                param_type2,
+                result_type,
+                visibility,
+            )
+        };
 
-        // Numeric operations are individually defined for each integer type in the impl language.
-        // The spec language uses the unified arbitrary precision num type instead.
-        use EntryVisibility::{Impl, Spec, SpecAndImpl};
-        use PA::BinOp_::*;
         let u8_ty = Type::Primitive(PrimitiveType::U8);
-        for prim_ty in [
-            PrimitiveType::U8,
-            PrimitiveType::U16,
-            PrimitiveType::U32,
-            PrimitiveType::U64,
-            PrimitiveType::U128,
-            PrimitiveType::U256,
-            PrimitiveType::Num,
-        ] {
-            let visibility = if prim_ty == PrimitiveType::Num {
-                Spec
-            } else {
-                Impl
-            };
-            let ty = Type::new_prim(prim_ty);
-            declare_bin(Add, Operation::Add, &ty, &ty, &ty, visibility);
-            declare_bin(Sub, Operation::Sub, &ty, &ty, &ty, visibility);
-            declare_bin(Mul, Operation::Mul, &ty, &ty, &ty, visibility);
-            declare_bin(Mod, Operation::Mod, &ty, &ty, &ty, visibility);
-            declare_bin(Div, Operation::Div, &ty, &ty, &ty, visibility);
-            declare_bin(BitOr, Operation::BitOr, &ty, &ty, &ty, visibility);
-            declare_bin(BitAnd, Operation::BitAnd, &ty, &ty, &ty, visibility);
-            declare_bin(Xor, Operation::Xor, &ty, &ty, &ty, visibility);
-            declare_bin(Shl, Operation::Shl, &ty, &u8_ty, &ty, visibility);
-            declare_bin(Shr, Operation::Shr, &ty, &u8_ty, &ty, visibility);
-            declare_bin(Lt, Operation::Lt, &ty, &ty, bool_t, visibility);
-            declare_bin(Le, Operation::Le, &ty, &ty, bool_t, visibility);
-            declare_bin(Gt, Operation::Gt, &ty, &ty, bool_t, visibility);
-            declare_bin(Ge, Operation::Ge, &ty, &ty, bool_t, visibility);
-        }
-
-        declare_bin(Range, Operation::Range, num_t, num_t, range_t, Spec);
-
-        declare_bin(Implies, Operation::Implies, bool_t, bool_t, bool_t, Spec);
-        declare_bin(Iff, Operation::Iff, bool_t, bool_t, bool_t, Spec);
-        declare_bin(And, Operation::And, bool_t, bool_t, bool_t, SpecAndImpl);
-        declare_bin(Or, Operation::Or, bool_t, bool_t, bool_t, SpecAndImpl);
+        let declare_arithm_ops = |trans: &mut ModelBuilder,
+                                  type_params: &[TypeParameter],
+                                  type_constraints: &BTreeMap<usize, Constraint>,
+                                  ty: Type,
+                                  visibility: EntryVisibility| {
+            declare_bin_gen(
```

Comment:
Could you use a loop to call declare_bin_gen for all of these more compactly?  (You need a loop over pairs.)

Reply:
Done.